crates/sui-indexer-alt/src/handlers/mod.rs

```diff
@@ -0,0 +1,428 @@
+
+                    metrics
+                        .total_committer_batches_attempted
+                        .with_label_values(&[H::NAME])
+                        .inc();
+
+                    metrics
+                        .committer_batch_size
+                        .with_label_values(&[H::NAME])
+                        .observe(batch.len() as f64);
+
+                    let guard = metrics
+                        .committer_commit_latency
+                        .with_label_values(&[H::NAME])
+                        .start_timer();
+
+                    match H::commit(&batch, &mut conn).await {
+                        Ok(affected) => {
+                            let elapsed = guard.stop_and_record();
+
+                            metrics
+                                .total_committer_rows_committed
+                                .with_label_values(&[H::NAME])
+                                .inc_by(batch.len() as u64);
+
+                            metrics
+                                .total_committer_rows_affected
+                                .with_label_values(&[H::NAME])
+                                .inc_by(affected as u64);
+
+                            debug!(
+                                pipeline = H::NAME,
+                                elapsed_ms = elapsed * 1000.0,
+                                attempt,
+                                affected,
+                                committed = batch.len(),
+                                pending = pending_rows,
+                                "Wrote batch",
+                            );
+
+                            pending_rows -= batch.len();
+                            attempt = 0;
+
+                            batch.clear();
+                            cool.reset();
+                        }
+
+                        Err(e) => {
+                            let elapsed = guard.stop_and_record();
+
+                            error!(
+                                pipeline = H::NAME,
+                                elapsed_ms = elapsed * 1000.0,
+                                attempt,
+                                committed = batch.len(),
+                                pending = pending_rows,
+                                "Error writing batch: {e}",
+                            );
+
+                            cool.reset_after(RETRY_INTERVAL);
+                            attempt += 1;
+                        }
+                    }
+
+                }
+
+                // If there are enough pending rows, and we've expended the cooldown, reset the
+                // commit polling interval so that on the next iteration of the loop, we will write
+                // out another batch.
+                _ = cool.tick(), if pending_rows > H::BATCH_SIZE => {
+                    poll.reset_immediately();
+                }
+
+                indexed = rx.recv(), if pending_rows < H::MAX_PENDING_SIZE => {
```

Comment:
At first, it was not obvious to me whether we'll ever get to this branch if `cool` and `poll` just keep ticking and we bias their branches, but upon another read I see that indeed this will be the only reachable branch when there's nothing in `pending` because the previous two branches are both conditional on `pending_rows` greater than some number.

Reply:
And I see now that `pending` is in fact the committer side buffering you mentioned in the comments for `COMMITTER_BUFFER`. So `MAX_PENDING_SIZE + COMMITTER_BUFFER` is basically the size of the channel between the handler and the committer. Is that right?

Reply:
I think the true buffer size is something like `MAX_PENDING_SIZE + H::FANOUT + COMMITTER_BUFFER`, but yeah, this is the right ball-park, and it is usually dominated by `MAX_PENDING_SIZE`, so we can treat this like `MAX_PENDING_SIZE + epsilon`.